xml/src/main/java/org/ehcache/xml/XmlConfiguration.java

```diff
@@ -103,58 +99,136 @@ public XmlConfiguration(URL url, final ClassLoader classLoader)
-    if(url == null) {
-      throw new NullPointerException("The url can not be null");
-    }
-    if(classLoader == null) {
-      throw new NullPointerException("The classLoader can not be null");
-    }
-    if(cacheClassLoaders == null) {
-      throw new NullPointerException("The cacheClassLoaders map can not be null");
-    }
-    this.xml = url;
+
+    this.source = requireNonNull(url, "The url can not be null");
+    requireNonNull(classLoader, "The classLoader can not be null");
+    this.cacheClassLoaders = requireNonNull(cacheClassLoaders, "The cacheClassLoaders map can not be null");
 
     try {
-      this.parser = new ConfigurationParser();
-      ConfigurationParser.XmlConfigurationWrapper xmlConfigurationWrapper = parser.parseConfiguration(url.toExternalForm(), classLoader, cacheClassLoaders);
-      configuration = xmlConfigurationWrapper.getConfiguration();
-      templates = xmlConfigurationWrapper.getTemplates();
-      try (BufferedReader buffer = new BufferedReader(new InputStreamReader(xml.openStream(), Charset.defaultCharset()))) {
-        xmlString = buffer.lines().collect(joining("\n"));
-      }
+      ConfigurationParser parser = new ConfigurationParser();
+      this.document = parser.uriToDocument(source.toURI());
+      ConfigurationParser.XmlConfigurationWrapper configWrapper = parser.documentToConfig(document, classLoader, cacheClassLoaders);
+      this.configuration = configWrapper.getConfiguration();
+      this.templates = configWrapper.getTemplates();
+
+      this.renderedDocument = ConfigurationParser.urlToText(url, document.getInputEncoding());
     } catch (XmlConfigurationException e) {
       throw e;
     } catch (Exception e) {
       throw new XmlConfigurationException("Error parsing XML configuration at " + url, e);
     }
   }
 
-  public XmlConfiguration(Configuration configuration) {
-    this.xml = null;
-    this.configuration = configuration;
-    this.templates = emptyMap();
+  /**
+   * Constructs an instance of XmlConfiguration from the given XML DOM.
+   *
+   * @param xml XML Document Object Model
+   *
+   * @throws XmlConfigurationException if anything went wrong parsing the XML
+   */
+  public XmlConfiguration(Document xml) throws XmlConfigurationException {
+    this(xml, ClassLoading.getDefaultClassLoader());
+  }
+
+  /**
+   * Constructs an instance of XmlConfiguration from the given XML DOM and using the provided {@code classLoader} to
+   * load user types (e.g. key and value Class instances).
+   *
+   * @param xml XML Document Object Model
+   * @param classLoader ClassLoader to use to load user types.
+   *
+   * @throws XmlConfigurationException if anything went wrong parsing the XML
+   */
+  public XmlConfiguration(Document xml, ClassLoader classLoader) throws XmlConfigurationException {
+    this(xml, classLoader, emptyMap());
+  }
+
+  /**
+   * Constructs an instance of XmlConfiguration from the given XML DOM and using the provided {@code classLoader} to
```

Comment:
Hmmm ... I presume the cacheClassLoaders are used to load user types for the indicated cache -- so, if I've provided cacheClassLoaders, what is classLoader used for?  If I have a ClassLoader specified for each cache, what should I specify for classLoader?

Reply:
Fixed... I think.